reuse.go

```diff
@@ -0,0 +1,187 @@
+package libp2pquic
+
+import (
+	"errors"
+	"net"
+	"sync"
+
+	srcs "github.com/lnykww/go-src-select"
+)
+
+type reuseConn struct {
+	net.PacketConn
+	mutex sync.Mutex
+	ref   int
+}
+
+func NewReuseConn(conn net.PacketConn) *reuseConn {
+	return &reuseConn{
+		PacketConn: conn,
+		ref:        1,
+	}
+}
+
+func (rc *reuseConn) Ref() bool {
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	if rc.ref == 0 {
+		return false
+	}
+	rc.ref++
+	return true
+}
+
+func (rc *reuseConn) Close() error {
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	var err error
+	switch rc.ref {
+	case 0: // cloesd, just return
+		return nil
+	case 1: // no reference, close the conn
+		err = rc.PacketConn.Close()
+	}
+	rc.ref--
+	return err
+}
+
+const RuseDialRetryTime = 3
+
+type Reuse struct {
+	mutex      sync.Mutex
+	unicast    map[string]map[int]*reuseConn
+	unspecific []*reuseConn
+	connGlobal *reuseConn
+}
+
+func NewReuse() *Reuse {
+	return &Reuse{
+		unicast:    make(map[string]map[int]*reuseConn),
```

Comment:
I assume the `int` should be the port number, right? This will need documentation here.